File: Ruby/merge_sort.rb
def merge_sort(fin)
  if fin.length <= 1
      return fin
  end

  mid = fin.length / 2
  lower = merge_sort(fin.slice(0, mid))
  upper = merge_sort(fin.slice(mid, fin.length - mid))

  fin = []
  tmp_a, tmp_b = 0, 0
  while tmp_a < lower.length && tmp_b < upper.length
      a, b = lower[tmp_a], upper[tmp_b]

      if a <= b
          fin << a
          tmp_a += 1
      else
          fin << b
          tmp_b += 1
      end
  end

  while tmp_a < lower.length
      fin << lower[tmp_a]
      tmp_a += 1
  end

  while tmp_b < upper.length
      fin << upper[tmp_b]
      tmp_b += 1
  end

  return fin
end

File: CPP/MergeSort.cpp
/* BCUR 2018 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
void mergeSort(vector<int> &fin, vector<int> &tmp, int low, int upper);
void merge(vector<int> &fin, vector<int> &tmp, int low, int mid, int upper);

void mergeSort(vector<int> &fin, vector<int> &tmp, int low, int upper) {
  if (low >= upper) {
    return;
  }
  int mid = (low + upper) / 2;
  mergeSort(fin, tmp, low, mid);
  mergeSort(fin, tmp, mid + 1, upper);
  merge(fin, tmp, low, mid, upper);
}

void merge(vector<int> &fin, vector<int> &tmp, int low, int mid, int upper) {
  int lowStart = low;
  int lowStop = mid;
  int upperStart = mid + 1;
  int upperStop = upper;
  int count = low;
  while (lowStart <= lowStop && upperStart <= upperStop) {
    if (fin[lowStart] < fin[upperStart]) {
      tmp[count++] = fin[lowStart++];
    } else {
      tmp[count++] = fin[upperStart++];
    }
  }
  while (lowStart <= lowStop) {
    tmp[count++] = fin[lowStart++];
  }
  while (upperStart <= upperStop) {
    tmp[count++] = fin[upperStart++];
  }
  for (int i = low; i <= upper; i++) {
    fin[i] = tmp[i];
  }
}
